File: mediation/mediatedviews/GoogleAdMob/ANAdAdapterBannerAdMob.m
/*   Copyright 2013 APPNEXUS INC
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "ANAdAdapterBannerAdMob.h"
#import "ANAdAdapterBaseDFP.h"
@interface ANAdAdapterBannerAdMob ()
@property (nonatomic, readwrite, strong) GADBannerView *bannerView;
@end


/**
 * Server side overridable
 */
@interface AdMobBannerServerSideParameters : NSObject
@property(nonatomic, readwrite) BOOL isSmartBanner;
@end
@implementation AdMobBannerServerSideParameters
@synthesize isSmartBanner;
@end





@implementation ANAdAdapterBannerAdMob
@synthesize delegate;
#pragma mark ANCustomAdapterBanner

- (void)requestBannerAdWithSize:(CGSize)size
             rootViewController:(nullable UIViewController *)rootViewController
                serverParameter:(nullable NSString *)parameterString
                       adUnitId:(nullable NSString *)idString
            targetingParameters:(nullable ANTargetingParameters *)targetingParameters
{
    ANLogDebug(@"Requesting AdMob banner with size: %fx%f", size.width, size.height);
	GADAdSize gadAdSize;
    
    AdMobBannerServerSideParameters *ssparam = [self parseServerSide:parameterString];
    
    // Allow server side to enable Smart Banners for this placement
    if (ssparam.isSmartBanner) {
        UIApplication *application = [UIApplication sharedApplication];
        BOOL orientationIsPortrait = UIInterfaceOrientationIsPortrait([application statusBarOrientation]);
        if(orientationIsPortrait) {
            gadAdSize = GADPortraitAnchoredAdaptiveBannerAdSizeWithWidth(size.width);
        } else {
            gadAdSize = GADLandscapeAnchoredAdaptiveBannerAdSizeWithWidth(size.height);
        }
    } else {
        gadAdSize = GADAdSizeFromCGSize(size);
    }
    self.bannerView = [[GADBannerView alloc] initWithAdSize:gadAdSize];
    
    self.bannerView.adUnitID = idString;
    self.bannerView.rootViewController = rootViewController;
    self.bannerView.delegate = self;
    [self.bannerView loadRequest:[self createRequestFromTargetingParameters:targetingParameters rootViewController: rootViewController]];
}

- (GADRequest *)createRequestFromTargetingParameters:(ANTargetingParameters *)targetingParameters rootViewController: (UIViewController *)rootViewController {
    return [ANAdAdapterBaseDFP googleAdMobRequestFromTargetingParameters:targetingParameters rootViewController: rootViewController];
}

- (AdMobBannerServerSideParameters*) parseServerSide:(NSString*) serverSideParameters
{
    AdMobBannerServerSideParameters *p = [AdMobBannerServerSideParameters new];
    NSError *jsonParsingError = nil;
    if (serverSideParameters == nil || [ serverSideParameters length] == 0) {
        return p;
    }
    NSData* data = [serverSideParameters dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary *jsonResponse = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonParsingError];
    
    if (jsonParsingError == nil && jsonResponse != nil) {
        p.isSmartBanner = [[jsonResponse valueForKey:@"smartbanner"] boolValue];
    }
    return p;
}


#pragma mark GADBannerViewDelegate

- (void)bannerViewDidReceiveAd:(nonnull GADBannerView *)bannerView
{
    ANLogDebug(@"AdMob banner did load");
    [self.delegate didLoadBannerAd:bannerView];
}
- (void)bannerView:(nonnull GADBannerView *)bannerView didFailToReceiveAdWithError:(nonnull NSError *)error
{
    ANLogDebug(@"AdMob banner failed to load with error: %@", error);
    [self.delegate didFailToLoadAd:[ANAdAdapterBaseDFP responseCodeFromRequestError:error]];
}

- (void)bannerViewWillPresentScreen:(nonnull GADBannerView *)bannerView{
    [self.delegate willPresentAd];
}

- (void)bannerViewWillDismissScreen:(nonnull GADBannerView *)bannerView {
    [self.delegate willCloseAd];
}

- (void)bannerViewDidDismissScreen:(nonnull GADBannerView *)bannerView {
    [self.delegate didCloseAd];
}

- (void)bannerViewDidRecordImpression:(GADBannerView *)bannerView {
    [self.delegate adDidLogImpression];
}

/* Tells the delegate that a click has been recorded for the ad. */
- (void)bannerViewDidRecordClick:(nonnull GADBannerView *)bannerView{
    [self.delegate adWasClicked];
}

- (void)dealloc
{
    ANLogDebug(@"AdMob banner being destroyed");
    self.bannerView.delegate = nil;
    self.bannerView = nil;
}


@end

File: examples/SimpleMediation/SimpleMediation/ANAdAdapterBannerAdMob.swift
//
//  CustomAdapter.swift
//  SimpleMediation
//
//  Created by System on 17/05/22.
//  Copyright © 2022 Xandr. All rights reserved.
//

import Foundation
import AppNexusSDK
import GoogleMobileAds

@objc(ANAdAdapterBannerAdMob)
public class ANAdAdapterBannerAdMob : NSObject , ANCustomAdapterBanner , GADBannerViewDelegate {
//    public var adDelegate: ANCustomAdapterBannerDelegate?

    
    public var delegate: AnyObject?
    var bannerView: GADBannerView!

    public func requestAd(with size: CGSize, rootViewController: UIViewController?, serverParameter parameterString: String?, adUnitId idString: String?, targetingParameters: ANTargetingParameters?) {
     
        bannerView = GADBannerView(adSize: GADAdSizeBanner)

        bannerView.adUnitID = "ca-app-pub-3940256099942544/2934735716"
        bannerView.rootViewController = rootViewController
        bannerView.delegate = self

        bannerView.load(GADRequest())

    }
    

      
    
    public func bannerViewDidReceiveAd(_ bannerView: GADBannerView) {
      print("bannerViewDidReceiveAd")
        self.delegate?.didLoadBannerAd(bannerView)
    }

    public func bannerView(_ bannerView: GADBannerView, didFailToReceiveAdWithError error: Error) {
      print("bannerView:didFailToReceiveAdWithError: \(error.localizedDescription)")
        self.delegate?.didFail(toLoadAd: ANAdResponseCode.unable_TO_FILL())

    }

    public func bannerViewDidRecordImpression(_ bannerView: GADBannerView) {
      print("bannerViewDidRecordImpression")
        self.delegate?.adDidLogImpression()

    }

    public func bannerViewWillPresentScreen(_ bannerView: GADBannerView) {
      print("bannerViewWillPresentScreen")
        self.delegate?.willCloseAd()

    }

    public func bannerViewWillDismissScreen(_ bannerView: GADBannerView) {
      print("bannerViewWillDIsmissScreen")
        self.delegate?.adDidLogImpression()

    }

    public func bannerViewDidDismissScreen(_ bannerView: GADBannerView) {
      print("bannerViewDidDismissScreen")
        self.delegate?.didCloseAd()

    }
    
}
